clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -491,6 +491,11 @@ static bool isSYCLUndefinedAllowed(const FunctionDecl *Callee,
   if (!Callee->getIdentifier())
     return false;
 
+  // Pure virtual functions need not be diagnosed here. If a definition
+  // is not provided in derived classes, clang will diagnose the call.
```

Comment:
Is the diagnostic about calling an undefined function without SYCL_EXTERNAL on it?

Reply:
Yes. That is the diagnostic we are suppressing for pure virtual functions now

Reply:
No, I mean if a definition is not provided in the derived classes.
In the example below, if Derived::testVF is not defined.

Reply:
Then we will get the normal clang diagnostic as seen here -  https://godbolt.org/z/dKMh9qheq

Reply:
This variant:
class PureVFClass {
  public:
  virtual void testVF() = 0;
};

class Derived : public PureVFClass {
   void testVF();
};

Reply:
This is UB as far as I know i.e. if a pure virtual function is used without being defined that is UB. I see Clang does not diagnose that. So after this change we won't either. Prior to this change the diagnostic about SYCL_EXTERNAL would have been emitted. 

The ask from SYCL spec authors was to match Clang to avoid the extra diagnostic about SYCL_EXTERNAL in some cases where we were (incorrectly?) diagnosing calls via pointer (E.g. https://godbolt.org/z/87cKqW3cz is not diagnosed in Clang but it diagnosed in deferred diagnostics) even if function was defined in a derived class. Thinking now about it, maybe I need to look into that further. If the function was defined in derived class why was the diagnostic was emitted in first place. Let me do some more testing and get back to you

Reply:
This is not UB.  In a standard compilation, there is no diagnostic is required for this case because the derived class virtual function could be defined in a different TU and linked together.   In the current SYCL model, we can't do that without indicating SYCL_EXTERNAL on the function.  I think it would be helpful to understand what exactly is being asked.

Reply:
`NPGKernel `does not overload the pure virtual function but IIRC classes that derive from it define it, which is why I had also assumed the same thing you did. I will take a closer look at update/close PR.

Reply:
I looked into this a bit further and see that we do emit a diagnostic pure virtual functions are called via pointer to Base class, even if there is a definition provided in Derived class. It is also interesting to note that clang does not throw an error in this case - https://godbolt.org/z/beYqo5e5a. While DPCPP throws the error about SYCL_EXTERNAL attribute.  This is because in DPCPP we visit call expressions for deferred diagnostics and we end up visiting the pure virtual function in very case even if the Base object points to derived class. Pasting AST we are traversing below. Note how the CXXMemberCallExpr refers to the Base class.

```
CXXMethodDecl 0xff72e90 <test.cpp:18:15, line:27:4> line:18:12 used constexpr operator() 'void () const' inline
  |-CompoundStmt 0xff77370 <col:17, line:27:4>
  | |-DeclStmt 0xff771a8 <line:24:7, col:16>
  | | `-VarDecl 0xff73008 <col:7, col:15> col:15 used d 'Derived':'Derived' callinit
  | |   `-CXXConstructExpr 0xff77180 <col:15> 'Derived':'Derived' 'void () noexcept'
  | |-DeclStmt 0xff77298 <line:25:7, col:28>
  | | `-VarDecl 0xff771d8 <col:7, col:27> col:20 used obj 'PureVFClass *' cinit
  | |   `-ImplicitCastExpr 0xff77278 <col:26, col:27> 'PureVFClass *' <DerivedToBase (PureVFClass)>
  | |     `-UnaryOperator 0xff77260 <col:26, col:27> 'Derived *' prefix '&' cannot overflow
  | |       `-DeclRefExpr 0xff77240 <col:27> 'Derived':'Derived' lvalue Var 0xff73008 'd' 'Derived':'Derived'
  | `-CXXMemberCallExpr 0xff77350 <line:26:7, col:21> 'void'
  |   `-MemberExpr 0xff77320 <col:7, col:14> '<bound member function type>' .testVF 0xff693b8
  |     `-ParenExpr 0xff77300 <col:7, col:12> 'PureVFClass':'PureVFClass' lvalue
  |       `-UnaryOperator 0xff772e8 <col:8, col:9> 'PureVFClass':'PureVFClass' lvalue prefix '*' cannot overflow
  |         `-ImplicitCastExpr 0xff772d0 <col:9> 'PureVFClass *' <LValueToRValue>
  |           `-DeclRefExpr 0xff772b0 <col:9> 'PureVFClass *' lvalue Var 0xff771d8 'obj' 'PureVFClass *'
```

I'm not sure if there is a way for us to tell if the pure virtual method is overridden when doing this traversal. Some quick things I tried did not work. I'll continue to look into it but if any of the reviewers  have a suggestion, I would greatly appreciate it.

Reply:
When we're processing the call through a base pointer, it really isn't possible to determine if a definition exists in a derived class. It would be very expensive to determine this for derived classes in the same TU and not possible to determine it for functions defined in a different TU. So my opinion here is to relax the diagnostic about undefined functions for virtual functions like this PR does.

This does cause a regression in the case Prem mentioned above i.e we will no longer have the diagnostic for undefined virtual function for this case -  

```
class PureVFClass {
public:
virtual void testVF() = 0;
};

class Derived : public PureVFClass {
void testVF();
};

void foo() {
   kernel<class test_kernel>(
      []() {
      PureVFClass* obj = &d;
      obj->.testVF();
   });
}

```

However, PR will make this case pass - 

```
class PureVFClass {
public:
virtual void testVF() = 0;
};

class Derived : public PureVFClass {
void testVF() { }
};

void foo() {
   kernel<class test_kernel>(
      []() {
      PureVFClass* obj = &d;
      obj->.testVF();
   });
}
```
IMO the second case is more important than the first if we want to enable virtual functions. Hopefully we can just count on the linker to handle the first case correctly.

Reply:
Unresolving this for now, to make sure that we finish the discussion about which diagnostics should be emitted and when. Tagging @gmlueck for awareness.

> When we're processing the call through a base pointer, it really isn't possible to determine if a definition exists in a derived class. It would be very expensive to determine this for derived classes in the same TU and not possible to determine it for functions defined in a different TU. So my opinion here is to relax the diagnostic about undefined functions for virtual functions like this PR does.

I'm okay with that proposal.

From what I understand, the diagnostic about calling an undefined function without `SYCL_EXTERNAL` is not even required by the spec, from [5.4. Language restrictions for device functions|https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:language.restrictions.kernels]:

> Any variable or function that is odr-used from a [device function](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#device-function) must be defined in the same translation unit as that use. However, a function may be defined in another translation unit if the implementation defines the `SYCL_EXTERNAL` macro as described in [Section 5.10.1](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#subsec:syclexternal).

Therefore, the whole diagnostic is solely a user experience question. If we don't have technical ability to reasonably implement the diagnostic in front-end, then there will be no such diagnostic for now in our implementation. Moreover, there are even request to turn current `error` diagnostic into a `warning` one: #1777.

Reply:
> Do we have a set of precise rules yet for calling virtual functions in device code? 

We don't have such set of rules defined at the moment.

> One possibility is this. When the static call tree contains a virtual function call, we look for all declarations in the current TU that could possibly be targets of that virtual call and consider each of these declared functions to be a "device function". In this case, the CFE should do normal "device function" error checking for those declared functions.

I believe that the fundamental feature of FE design is that there is no "for all declarations". You have current declaration context and a declaration you are looking at and unlikely anything more than that. In particular, you may have not yet seen  the whole file and simply don't know all the possible candidates.

Reply:
>>One possibility is this. When the static call tree contains a virtual function call, we look for all declarations in the current TU that could possibly be targets of that virtual call and consider each of these declared functions to be a "device function". In this case, the CFE should do normal "device function" error checking for those declared functions.

In order to do this, we'll have to search the entire AST for all types,  check if it derives from the base class and then search it's functions to see if defines the virtual functions. This is not impossible to do but it is expensive. I do not know if upstream clang would ever accept this. Tagging @erichkeane for his thoughts on it.

Reply:
> > Do we have a set of precise rules yet for calling virtual functions in device code?
> 
> We don't have such set of rules defined at the moment.

I think we need to formalize this set of rules soon because it will drive some design decisions in the CFE.

> In order to do this, we'll have to search the entire AST for all types, check if it derives from the base class and then search it's functions to see if defines the virtual functions.

Won't we need to do this anyways?  How else will these functions be emitted into the LLVM IR, which is eventually used to construct the device code image?

Reply:
So the problem is that the 'must be defined in a way that the current TU can reach it' rule is one we can implement; just not with virtual functions.  Thanks to the way inheritance works, we have no way of knowing in the case of a virtual function call, WHICH derived class is being called, it could be one defined in the current TU, or one from an entirely different TU.  

And I cannot come up with a reasonable rule for how we can do that sort of check (or even another check that makes it possible).  My only solution is to do away with the 'must be defined in a way that the current TU can reach it', but that breaks a number of other things as well.

Reply:
I think this all comes down to the same core problem: what is the precise rule for identifying "device functions" in the presence of virtual function calls.

I can think of a couple possibilities:

1. We require the application to add a decoration to any member function that could be the target of a virtual function call in order to include it as a "device function".

2. The compiler automatically identifies member functions that could possibly be the target of a virtual function call and implicitly considers these as "device functions".  Obviously, the compiler can only do this for member functions that are declared in the same TU as the virtual function call site.  If the user wants a member function in another TU to be consider a "device function", they will need to add some annotation.

From what I recall, we thought customers would not be happy with (1) because they would need to add annotations even in common cases.  @AlexeySachkov may recall more details, though.

@erichkeane: Are you saying that it's not possible to do (2), or are you merely saying that it's not possible to identify the function that is definitely called from a virtual call site?

Reply:
We can't really do #2, since we don't know how that type was passed around.  Any virtual function ends up being potentially called on the device.  Even #1 we couldn't really enforce besides it failing linking (which, is exactly why C/C++ work this way).

Reply:
> Even (1) we couldn't really enforce besides it failing linking (which, is exactly why C/C++ work this way).

I was not proposing that the CFE enforce this.  Actually, it probably wouldn't trigger a link error either.  I'm not sure we can even detect this at the time the kernel is launched, so it might just be UB if the user forgot the decoration.

> We can't really do (2), since we don't know how that type was passed around. Any virtual function ends up being potentially called on the device.

I think I'm suggesting something much simpler than you are thinking.  I'm suggesting that the CFE do the following for each virtual function call in device code:

* Let's refer to the class name of the call as _Base_ and the function name as _Func_.
* Search for all classes defined in the current TU that derive from _Base_ and declare _Func_.  Let's call these functions _Derived_::_Func_.
* The CFE implicitly assumes that _Derived_::_Func_ is a device function, even if there is no direct call from device code.

Reply:
Ah, so the idea here is that any type used device side has all of its virtual functions emitted, PLUS any detectable types derived from that?  We'd end up emitting a lot of additional functions that way.  The reason we'd have to do ALL virtual functions in `Base` (even if not called) is because:

```
struct Base {
virtual void foo() = 0;
virtual void bar();
virtual void baz(){}
};
SYCL_EXTERNAL void External(Base *B);

struct Derived : Base {
virtual void foo(){}
};
void OtherFunction() {
  Derived D;
  External(&D); // External might call foo, bar, or baz!
}

```

Additionally, we'd have to have any virtual function defined at all, since the TU containing 'Base::bar` might not use `Base` in a function, right?  What if the place that has `External` is the one that uses it?  Or a 3rd TU that has device code?  

DO we have to emit every variant of every virtual function that we run into?  With a function like `baz`, does that cause us to emit that function a ton of times as a result?

Reply:
I don't think this example demonstrates a problem with my proposal:

> External(&D); // External might call foo, bar, or baz!

This is OK.  Let's call the TU you show above "TU1".  Let's call the TU that defines `External` "TU2".  The declaration of `External` in TU2 is required to have `SYCL_EXTERNAL`, so the body of `External` will be in the device code section of TU2.  If the body of `External` calls `baz`, the definition of that function will also be recognized as device code and will also be included in TU2.  If `External` calls `bar`, the compiler will diagnose an error because `bar` is not declared with `SYCL_EXTERNAL` (unless TU2 also happens to define `bar`).  If `External` calls any of `foo`, `bar`, or `baz`, the definitions of those functions in any derived classes in TU2 are also considered device code.

> Additionally, we'd have to have any virtual function defined at all, since the TU containing `Base::bar` might not use `Base` in a function, right?

If any device code in any TU calls `Base::bar`, then the TU defining `Base::bar` is required to declare that function as `SYCL_EXTERNAL`.  Therefore, `Base::bar` will be included as device code.  Note that this has nothing to do with virtual functions.  This is how `SYCL_EXTERNAL` is specified for any function that is called from device code.

Reply:
> Ah, so the idea here is that any type used device side has all of its virtual functions emitted, PLUS any detectable types derived from that? We'd end up emitting a lot of additional functions that way. The reason we'd have to do ALL virtual functions in Base (even if not called)

@erichkeane , that is what we currently do for virtual functions: we emit them **all** into device code. I suppose that the reason for that is `vtable`: since we don't know (in general) where else our virtual method could be called, we have to emit them all. They have external linkage by default after all.

`SYCL_EXTERNAL` looks weird in that context: it kind of says that default linkage of device functions is `internal` and you have to specify the macro to make it `external`, whilst it works the opposite way in C++.

From customer point of view (especially if their background is plain C++), I think it would be best **not** to require any special markup on any virtual function **declaration** and treat calls to undefined methods and pure virtual functions as a UB.

> I think this all comes down to the same core problem: what is the precise rule for identifying "device functions" in the presence of virtual function calls.
>
> 1. We require the application to add a decoration to any member function that could be the target of a virtual function call in order to include it as a "device function".
> 
> From what I recall, we thought customers would not be happy with (1) because they would need to add annotations even in common cases. @AlexeySachkov may recall more details, though.

I'm not sure if this can be solved without having user-provided input in the source code. Let's consider the following example:

```c++
class Base {
public:
  virtual void foo() = 0;
};
class DerivedForDevice : public Base {
public:
  void foo() override {
    // legal device code
  }
};
class DerivedForHost : public Base {
public:
  void foo() override {
    new Object();
    throw runtime_error(42);
  }
};

SYCL_EXTERNAL void bar(Base *obj) {
  obj->foo();
}
```

Compiler has no idea which particular object will be passed to `bar`. We should either say that if at least one override of a member function within a whole inheritance graph is device function, then all other overrides of the same member function (including its original definition) are also considered to be device functions. I suppose that is exactly what @gmlueck suggests. That would make example below illegal. I'm not sure how close it is to someone's real life use-case, though.

An alternative approach would be: require customer to put a special attribute/macro on virtual function **definitions** (we could re-use existing `[[intel::device_indirectly_callable]]`), which are supposed to be device callable and **not** emit any "undefined function without `SYCL_EXTERNAL`" diagnostics for virtual functions. User-provided markup will help with:
- filtering-out host-only virtual functions. Modifying `vtable` in FE could be complicated and potentially may break object layout between host and device (and won't be an option in one-pass compiler). Instead, we could use the attribute to later reset some `vtable` elements to `nullptr` at LLVM IR level
- ensuring that a virtual function will be emitted into device code in case it is defined outside of a class body, in a separate translation unit without any uses of it

It just seems to me that the diagnostic with `SYCL_EXTERNAL` for virtual functions can't be reliably implemented. Even in plain C++ there won't be even a link error if virtual function definition is missing, because it is referenced indirectly. Therefore, I don't see how we can come up with such diagnostic for SYCL and why do we even need that (considering aim to align with C++ and eventually be a part of it).

Reply:
>>If any device code in any TU calls Base::bar, then the TU defining Base::bar is required to declare that function as SYCL_EXTERNAL. Therefore, Base::bar will be included as device code. Note that this has nothing to do with virtual functions. This is how SYCL_EXTERNAL is specified for any function that is called from device code.

Isn't the distinction with virtual functions important though? When you call a function using a base pointer pointing to a derived object, it calls the base method unless the method is virtual. For the former, you just need to check if the base type. For the latter you have to check every derived type and emit all virtual functions like @erichkeane mentioned above. 

>>@erichkeane , that is what we currently do for virtual functions: we emit them all into device code. I suppose that the reason for that is vtable: since we don't know (in general) where else our virtual method could be called, we have to emit them all. They have external linkage by default after all.

@AlexeySachkov is this confirmed? The last time we spoke about this, there was some doubt about this right?

Reply:
> Isn't the distinction with virtual functions important though? When you call a function using a base pointer pointing to a derived object, it calls the base method unless the method is virtual. For the former, you just need to check if the base type. For the latter you have to check every derived type and emit all virtual functions like @erichkeane mentioned above.

The question about whether the CFE needs to iterate over all the base classes seems unrelated to `SYCL_EXTERNAL`.  Even when there is a call to `Base::baz` (which is defined in the current TU), the CFE may need to iterate over all the derived classes to find other definitions of `baz`.

It all comes down to the rules we put in the extension specification for virtual function support.  Does the user need to add some annotation to virtual function definitions in order for them to be considered "device code", or does the compiler identify some subset of virtual functions that could potentially be called and automatically consider them "device code"?

@AlexeySachkov, I think this is your feature, so you should be the one who documents these rules.  I'm happy to consult if you want to talk through the issues, and I'm open to any reasonable set of rules.  I do think we need a precise set of rules, though, because users need to know how to write code that is guaranteed to run correctly.  If we expect other vendors to implement our virtual function extension, we also need precise rules to guarantee portability across vendors.

> It just seems to me that the diagnostic with SYCL_EXTERNAL for virtual functions can't be reliably implemented.

Let's work out the rules for determining which virtual functions are "device code" first, and then worry about the interaction with `SYCL_EXTERNAL`.  I think the interaction will be much clearer once we understand the rules.

Reply:
> @AlexeySachkov mentioned that we emit all virtual functions in DPCPP compiler. If that's the case, it means we've modified 'normal' clang behavior somewhere already (possibly because we want all definitions available as possible device code?), but again I do not know the specifics here.

@elizabethandrews, I will do some additional experiments, but so far my estimate of "all" was based on an example we have. It seems to me that the behavior is the following here:
- `clang` tries to de-virtualize as much as possible and avoid emitting any unnecessary code (as in the example below). I don't think that we have modified that logic and I don't think that we should do that. If `clang` says that particular virtual function is not needed for some reason, we shouldn't change that, since we are not trying to bring any new rules into C++ around virtual functions. The only thing we need to decide on is which virtual functions should be considered as device functions and at which stage
- if an object of a class is used in kernel code, _all_ virtual functions of that class (as well as base and derived one), are now considered to be device functions. Note that they are emitted into device code, but not necessarily checked if they comply with device code restrictions. My guess is that happens, because those functions are part of the vtable